File: ValidSudoku/valid_sudoku.py
def is_valid_sudoku(board: list[list[str]]) -> bool:
    column_sets = [set() for i in range(9)]
    square_sets = [set() for i in range(9)]
    row_set = set()
    for m, row in enumerate(board):
        for n, cell in enumerate(row):
            if not cell.isdigit():
                continue
            i = n // 3 + (3 * (m // 3))
            d = int(cell)
            col_set = column_sets[n]
            sqr_set = square_sets[i]
            if d in row_set or d in col_set or d in sqr_set:
                return False
            row_set.add(d)
            col_set.add(d)
            sqr_set.add(d)
        row_set = set()
    return True

# [["5","3",".",".","7",".",".",".","."] ,["6",".",".","1","9","5",".",".","."] ,[".","9","8",".",".",".",".","6","."] ,["8",".",".",".","6",".",".",".","3"] ,["4",".",".","8",".","3",".",".","1"] ,["7",".",".",".","2",".",".",".","6"] ,[".","6",".",".",".",".","2","8","."] ,[".",".",".","4","1","9",".",".","5"] ,[".",".",".",".","8",".",".","7","9"]]

File: ValidSudoku/ValidSudoku.hs
import Data.Char (isDigit)
import Data.Set qualified as S

type IntSet = S.Set Int

parseCell :: Char -> Int
parseCell c
  | isDigit c = read [c]
  | otherwise = 0

replace :: Int -> a -> [a] -> [a]
replace i a l
  | i + 1 > length l = error "Index error"
  | otherwise = take i l ++ (a : drop (i + 1) l)

validSudoku :: [String] -> Bool
validSudoku = go (0, 0) S.empty columnSets squareSets
  where
    columnSets = [S.empty | _ <- [1 .. 9]] :: [IntSet]
    squareSets = [S.empty | _ <- [1 .. 9]] :: [IntSet]
    go :: (Int, Int) -> IntSet -> [IntSet] -> [IntSet] -> [String] -> Bool
    go _ _ _ _ [] = True
    go (m, n) s colSets sqrSets ([] : rs) = go (m + 1, 0) S.empty colSets sqrSets rs
    go (m, n) s colSets sqrSets ((cell : cells) : rs)
      | d == 0 = go (m, n + 1) s colSets sqrSets (cells : rs)
      | S.member d s = False
      | S.member d (colSets !! n) = False
      | S.member d (sqrSets !! i) = False
      | otherwise = go (m, n + 1) (S.insert d s) colSets' sqrSets' (cells : rs)
      where
        d = parseCell cell
        i = div n 3 + (3 * div m 3)
        colSet = S.insert d $ colSets !! n
        sqrSet = S.insert d $ sqrSets !! i
        colSets' = replace n colSet colSets
        sqrSets' = replace i sqrSet sqrSets
